core/src/main/java/com/datastrato/gravitino/lock/TreeLockNode.java

```diff
@@ -0,0 +1,170 @@
+   */
+  void unlock(LockType lockType) {
+    if (lockType == LockType.READ) {
+      readWriteLock.readLock().unlock();
+    } else {
+      readWriteLock.writeLock().unlock();
+    }
+
+    LOG.trace("Node {} has been unlock with '{}' lock", name, lockType);
+  }
+
+  /**
+   * Get the tree lock node by the given name. If the node doesn't exist, create a new TreeNode.
+   *
+   * <p>Note: This method should always be guarded by object lock.
+   *
+   * @param name The name of a resource such as entity or others.
+   * @return The tree lock node of this ident.
+   */
+  TreeLockNode getOrCreateChild(String name) {
+    TreeLockNode childNode =
+        childMap.computeIfAbsent(
+            name,
+            k -> {
+              // If the total node count is greater than the max node count, in case of memory
+              // leak and explosion, we should throw an exception.
+              long currentNodeCount = lockManager.totalNodeCount.get();
+              if (currentNodeCount > lockManager.maxTreeNodeInMemory) {
+                throw new IllegalStateException(
+                    "The total node count '"
+                        + currentNodeCount
+                        + "' has reached the max node count '"
+                        + lockManager.maxTreeNodeInMemory
+                        + "', please increase the max node count or wait for a while to avoid the performance issue.");
+              }
+
+              TreeLockNode newNode = new TreeLockNode(name, lockManager);
+              lockManager.totalNodeCount.getAndIncrement();
+              LOG.trace("Create tree lock node '{}' as a child of '{}'", name, this.name);
+              return newNode;
+            });
+
+    childNode.addReference();
+    return childNode;
+  }
+
+  /**
+   * Get all the children of this node. The returned list is unmodifiable and the order is random.
+   * The reason why we return a random order list is that we want to avoid the cases that the first
+   * child is always to be dropped firstly and the last child is always the last to be dropped,
+   * chances are that the first child is always the most popular one and the last child is always
+   * the least popular.
+   *
+   * @return The list of all the children of this node.
+   */
+  synchronized List<TreeLockNode> getAllChildren() {
+    List<TreeLockNode> children = Lists.newArrayList(childMap.values());
+    Collections.shuffle(children);
+    return Collections.unmodifiableList(children);
+  }
+
+  /**
+   * Remove the child node by the given name identifier.
+   *
+   * <p>Note: This method should be guarded by object lock.
+   *
+   * @param name The name of a resource such as entity or others.
+   */
+  void removeChild(String name) {
+    childMap.remove(name);
+  }
+
+  @Override
+  public String toString() {
+    final StringBuilder sb = new StringBuilder("TreeLockNode{");
+    sb.append("ident=").append(name);
+    sb.append('}');
+    return sb.toString();
```

Comment:
Include the hash code for debugging purposes.

Reply:
OK